File: pharmacie/src/main/java/io/pharmacie/FilterVille.java
package io.pharmacie;

import android.app.ProgressDialog;
import android.content.Intent;
import android.support.annotation.NonNull;
import android.support.design.widget.BottomNavigationView;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.KeyEvent;
import android.view.MenuItem;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Filter;
import android.widget.ListView;
import android.widget.Spinner;
import android.widget.TextView;
import android.widget.Toast;
import java.util.ArrayList;
import java.util.List;
import io.pharmacie.Retrofit.Api;
import io.pharmacie.models.Commune;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

import static io.pharmacie.R.string.loading;


public class FilterVille extends AppCompatActivity {

    private static final int REQUEST_SIGNUP = 0;
    ListView _myListView;
    Spinner _mySpinner;
    ArrayList<Commune> communes = new ArrayList<>();
    public static String commune;
    //List<Commune> ;
    String[] wilaya = {"All","Adrar","Chlef","Laghouat","Oum El Bouaghi","Batna","Bejaia","Biskra","Bechar","Blida","Bouira","Tamanrasset",
            "Tebéssa","Tlemcen","Tiaret","Tizi Ouzou","Alger","Djelfa","Jijel","Sétif","Saida","Skikda","Sidi Bel Abbès","Annaba","Guelma","Constantine"
            ,"Mèdéa","Mostaganem","Msila","Mascara","Ouargla","Oran","El Bayadh","Illizi","Bourdj Bou Arreridj","Boumerdès","Tarf","Tindouf","Tissemsilt","El Oued"
            ,"Khenschla","Souk Ahras","Tipaza","Mila","Ain Defla","Naama","Ain Tèmouchent","Ghardaia","Relizane"};
    ArrayAdapter<Commune> adapter;

    @Override
    protected void onCreate(Bundle savedInstanceState){
        super.onCreate(savedInstanceState);
        setContentView(R.layout.choosecity_layout);

        View decorView = getWindow().getDecorView();
        MainActivity.hideSystemUI(decorView);
        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl(Api.Base_Url)
                .addConverterFactory(GsonConverterFactory.create())
                .build();

        Api api = retrofit.create(Api.class);

        ProgressDialog dialog=new ProgressDialog(FilterVille.this);
        dialog.setMessage(getResources().getString(R.string.loading));
        dialog.setCancelable(false);
        dialog.setInverseBackgroundForced(false);
        dialog.show();

        Call<List<Commune>> call = api.getCommunes();
        call.enqueue(new Callback<List<Commune>>() {
            @Override
            public void onResponse(Call<List<Commune>> call, Response<List<Commune>> response) {
                List<Commune> coms = response.body();

                for(Commune c : coms){
                    communes.add(c);
                }
                initializeView();
                dialog.hide();
            }

            @Override
            public void onFailure(Call<List<Commune>> call, Throwable t) {
                Toast.makeText(getApplicationContext(),t.getMessage(),Toast.LENGTH_SHORT).show();
                dialog.hide();
            }
        });
    }

    private void initializeView(){
        _mySpinner = findViewById(R.id.mySpinner);
        _mySpinner.setAdapter(new ArrayAdapter<>(FilterVille.this,android.R.layout.simple_list_item_1,wilaya));
        _myListView = findViewById(R.id.myListView);
        _myListView.setAdapter(new ArrayAdapter<>(FilterVille.this,android.R.layout.simple_list_item_1,communes));
        _mySpinner.setOnItemSelectedListener (new AdapterView.OnItemSelectedListener(){

            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                if(position>=0 && position < wilaya.length){
                    getSelectectWilayaData(position);
                }else{
                    Toast.makeText(FilterVille.this,"Selected Wilaya Doesn't exist",Toast.LENGTH_SHORT).show();
                }
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {

            }
        });

        _myListView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position,
                                    long id) {
                commune = ((TextView) view).getText().toString().replaceAll("\\s", "%20");
                MainActivity.PreviousClass = FilterVille.class;
                Intent intent = new Intent(getApplicationContext(),showPharmacies.class);
                startActivity(intent);
                finish();
            }
        });
    }

    private void getSelectectWilayaData(int wilayacode) {
        ArrayList<Commune> coms = new ArrayList<>();
        if(wilayacode == 0){
            adapter= new ArrayAdapter<>(FilterVille.this,android.R.layout.simple_list_item_1,communes);
        }else{
            for(Commune commune:communes){
                if(commune.getWiyalaCode() == wilayacode){
                    coms.add(commune);
                }
            }
            adapter= new ArrayAdapter<>(FilterVille.this,android.R.layout.simple_list_item_1,coms);
        }
        _myListView.setAdapter(adapter);
    }
    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event)
    {
        if (keyCode == KeyEvent.KEYCODE_BACK)
        {
            Intent intentprevious = new Intent(getApplicationContext(),MainActivity.class);
            startActivity(intentprevious);
            finish();
            overridePendingTransition(R.anim.push_left_in, R.anim.push_left_out);
        }

        if (keyCode == KeyEvent.KEYCODE_HOME)
        {

        }

        // // TODO Auto-generated method stub
        return super.onKeyDown(keyCode, event);

    }
}

File: pharmacie/src/main/java/io/pharmacie/FilterVille.kt
package io.pharmacie

import android.app.ProgressDialog
import android.content.Context
import android.content.Intent
import android.net.ConnectivityManager
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import android.view.KeyEvent
import android.view.View
import android.widget.*
import androidx.core.view.isVisible
import androidx.recyclerview.widget.LinearLayoutManager
import java.util.ArrayList
import io.pharmacie.Retrofit.Api
import io.pharmacie.models.Commune
import io.pharmacie.models.RoomDatabase.DataBase_Instance
import io.pharmacie.models.RoomDatabase.RepoPharmacie_Local
import io.pharmacie.models.pharmacy
import kotlinx.android.synthetic.main.activity_ville.*
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.util.Observer


class FilterVille : AppCompatActivity() {
    lateinit var _myListView: ListView
     lateinit var _mySpinner: Spinner
    lateinit var _modeoffline: ListView

      var communes = ArrayList<String>()
    var liscommune=ArrayList<Commune>()
    var array=ArrayList<String>()

    //List<Commune> ;

      var wilaya = arrayOf("TOUS LES WILAYA", "Adrar", "Chlef", "Laghouat", "Oum El Bouaghi", "Batna", "Bejaia", "Biskra", "Bechar", "Blida", "Bouira", "Tamanrasset", "Tebéssa", "Tlemcen", "Tiaret", "Tizi Ouzou", "Alger", "Djelfa", "Jijel", "Sétif", "Saida", "Skikda", "Sidi Bel Abbès", "Annaba", "Guelma", "Constantine", "Mèdéa", "Mostaganem", "Msila", "Mascara", "Ouargla", "Oran", "El Bayadh", "Illizi", "Bourdj Bou Arreridj", "Boumerdès", "Tarf", "Tindouf", "Tissemsilt", "El Oued", "Khenschla", "Souk Ahras", "Tipaza", "Mila", "Ain Defla", "Naama", "Ain Tèmouchent", "Ghardaia", "Relizane")
     lateinit var adapter:ArrayAdapter<String>

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.choosecity_layout)

        val decorView = window.decorView
        MainActivity.hideSystemUI(decorView)
        if (verifyAvailableNetwork(AppCompatActivity())) {
            val retrofit = Retrofit.Builder()
                .baseUrl(Api.Base_Url)
                .addConverterFactory(GsonConverterFactory.create())
                .build()

            val api = retrofit.create<Api>(Api::class.java!!)

            /* val dialog = ProgressDialog(this@FilterVille)
        dialog.setMessage(resources.getString(R.string.loading))
        dialog.setCancelable(false)
        dialog.setInverseBackgroundForced(false)
        dialog.show()*/

            val call = api.communes()
            call.enqueue(object : Callback<List<Commune>> {
                override fun onResponse(call: Call<List<Commune>>, response: Response<List<Commune>>) {
                    val coms = response.body()
                    liscommune = coms as ArrayList<Commune>
                    for (c in coms!!) {

                        Log.e("MUSTAPHADEBBIH", c.codeWilaya.toString())

                        communes.add(c.nomCommune.toString())

                    }
                    initializeView()
                    //dialog.hide()
                }

                override fun onFailure(call: Call<List<Commune>>, t: Throwable) {
                    Toast.makeText(applicationContext, t.message, Toast.LENGTH_SHORT).show()
                    //dialog.hide()
                }
            })
        }else{
            mySpinner.setVisibility(View.GONE)
            //myListView.setVisibility(View.GONE)
            // modeoffline!!.setVisibility(View.VISIBLE)
            Log.e("dataromlist", "herrrrrrrrrrrrre")
            val RepoLocal= RepoPharmacie_Local(this.applicationContext)
            RepoLocal.getAllPharmacie().observe(this, androidx.lifecycle.Observer{

                    val list=it
                    var nomPharma=ArrayList<String>()
                    for (c in list!!) {

                        Log.e("MUSTAPHIH", c.nomPrenomPharmacien.toString())
                        nomPharma.add(c.nomPrenomPharmacien)


                    }
                   // _modeoffline = findViewById(R.id.modeoffline!!)
                    _myListView = findViewById(R.id.myListView)
                    _myListView!!.adapter = ArrayAdapter(this@FilterVille, android.R.layout.simple_list_item_1, nomPharma)

                

            })

        }
    }

    private fun initializeView() {
        _mySpinner = findViewById(R.id.mySpinner)
        _mySpinner.adapter = ArrayAdapter(this@FilterVille, android.R.layout.simple_list_item_1, wilaya)
        _myListView = findViewById(R.id.myListView)


        Log.e("mustaphaamine",array.toString())
        _myListView.adapter = ArrayAdapter(this@FilterVille, android.R.layout.simple_list_item_1, communes)

        _mySpinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {

            override fun onItemSelected(parent: AdapterView<*>, view: View, position: Int, id: Long) {
                if (position >= 0 && position < wilaya.size) {
                    getSelectectWilayaData(position)
                } else {
                    Toast.makeText(this@FilterVille, "Selected Wilaya Doesn't exist", Toast.LENGTH_SHORT).show()
                }
            }

            override fun onNothingSelected(parent: AdapterView<*>) {

            }
        }

        _myListView.onItemClickListener = AdapterView.OnItemClickListener { parent, view, position, id ->
            commune = (view as TextView).text.toString().replace("\\s".toRegex(), "%20")
            MainActivity.PreviousClass = FilterVille::class.java
            val intent = Intent(applicationContext, showPharmacies::class.java)
            startActivity(intent)
            finish()
        }
    }

    private fun getSelectectWilayaData(wilayacode: Int) {
        var coms = ArrayList<String>()
        if (wilayacode == 0) {
            adapter = ArrayAdapter(this@FilterVille, android.R.layout.simple_list_item_1, communes)
        }else {
            Log.e("wilaya",wilayacode.toString())
            Log.e("listcommun",liscommune.toString())
            for (commune in liscommune) {
                Log.e("listcommun",commune.nomCommune)
                Log.e("listcommuncode",commune.codeWilaya.toString())
                if (commune.codeWilaya == wilayacode) {
                    Log.e("listcommun",commune.nomCommune)
                    coms.add(commune.nomCommune.toString())
                }
            }
            adapter = ArrayAdapter(this@FilterVille, android.R.layout.simple_list_item_1, coms)
        }
        _myListView.adapter = adapter
    }

    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {
        if (keyCode == KeyEvent.KEYCODE_BACK) {
            val intentprevious = Intent(applicationContext, MainActivity::class.java)
            startActivity(intentprevious)
            finish()
            overridePendingTransition(R.anim.push_left_in, R.anim.push_left_out)
        }

        if (keyCode == KeyEvent.KEYCODE_HOME) {

        }

        // // TODO Auto-generated method stub
        return super.onKeyDown(keyCode, event)

    }
    fun verifyAvailableNetwork(activity:AppCompatActivity):Boolean{
        val connectivityManager=getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        val networkInfo=connectivityManager.activeNetworkInfo
        return  networkInfo!=null && networkInfo.isConnected
    }
    companion object {

        private val REQUEST_SIGNUP = 0
        lateinit var commune: String


    }
}
